<!DOCTYPE html>
<html>
  <head>
    <title>Sequences In Bioconductor</title>
    <meta charset="utf-8">
    <meta name="author" content="Rockefeller University, Bioinformatics Resource Centre" />
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/default-fonts.css" rel="stylesheet" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Sequences In Bioconductor
### Rockefeller University, Bioinformatics Resource Centre
### <a href="http://rockefelleruniversity.github.io/SequencesInBioconductor/" class="uri">http://rockefelleruniversity.github.io/SequencesInBioconductor/</a>

---





# Genomic Sequences

As we have seen earlier, genomic sequences are often stored in FASTA format.

&lt;div align="center"&gt;
&lt;img src="imgs/fasta.png" alt="igv" height="400" width="400"&gt;
&lt;/div&gt;


---

# Genomic Sequences in Bioconductor.

Genomic sequences can be handled in Bioconductor using the functions in the **Biostrings** package and full genome sequences for many model organisms are contained within the **BSgenome** packages.

---

# BSgenome packages.

To make use of a BSgenome package we must first install and load the library.

The resulting BSgenome object (here BSgenome.Mmusculus.UCSC.mm10) contains the full genome sequence for the model organism.


```r
library(BSgenome.Mmusculus.UCSC.mm10)
class(BSgenome.Mmusculus.UCSC.mm10)
```


```
## [1] "BSgenome"
## attr(,"package")
## [1] "BSgenome"
```
---


```r
BSgenome.Mmusculus.UCSC.mm10
```

```
## Mouse genome:
## # organism: Mus musculus (Mouse)
## # provider: UCSC
## # provider version: mm10
## # release date: Dec. 2011
## # release name: Genome Reference Consortium GRCm38
## # 66 sequences:
## #   chr1                 chr2                 chr3                
## #   chr4                 chr5                 chr6                
## #   chr7                 chr8                 chr9                
## #   chr10                chr11                chr12               
## #   chr13                chr14                chr15               
## #   ...                  ...                  ...                 
## #   chrUn_GL456372       chrUn_GL456378       chrUn_GL456379      
## #   chrUn_GL456381       chrUn_GL456382       chrUn_GL456383      
## #   chrUn_GL456385       chrUn_GL456387       chrUn_GL456389      
## #   chrUn_GL456390       chrUn_GL456392       chrUn_GL456393      
## #   chrUn_GL456394       chrUn_GL456396       chrUn_JH584304      
## # (use 'seqnames()' to see all the sequence names, use the '$' or '[['
## # operator to access a given sequence)
```
---
# Data in BSgenome objects.

Simply by typing object in we have retrieved some important information.

Importantly it tells us how we can access information in the object.

- **seqnames()** function to retrieve all contig names.
- **$** and **[[]]** accessors to retrieve a given contig's sequence.
---
# Contig names.

We can retrieve contig names using **seqnames()** function.


```r
contigNames &lt;- seqnames(BSgenome.Mmusculus.UCSC.mm10)
contigNames[1:22]
```

```
##  [1] "chr1"  "chr2"  "chr3"  "chr4"  "chr5"  "chr6"  "chr7"  "chr8" 
##  [9] "chr9"  "chr10" "chr11" "chr12" "chr13" "chr14" "chr15" "chr16"
## [17] "chr17" "chr18" "chr19" "chrX"  "chrY"  "chrM"
```
---

# Contig lengths

We can also retrieve contig lengths using **seqlengths()** function.


```r
contigLengths &lt;- seqlengths(BSgenome.Mmusculus.UCSC.mm10)
contigLengths[1:4]
```

```
##      chr1      chr2      chr3      chr4 
## 195471971 182113224 160039680 156508116
```

---

# Retreiving sequence information.

Now we know sequence names, we can retrieve sequence information using either the **$** or **[[]]** accessors.

Here we use the data.frame and list accessor **$** with name of contig of interest

```r
chr19_Seq &lt;- BSgenome.Mmusculus.UCSC.mm10$chr19
chr19_Seq
```

```
##   61431566-letter "DNAString" instance
## seq: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
```
---
# Retreiving sequence information.


now we use list accessor **[[]]** with name of contig of interest

```r
chr19_Seq &lt;- BSgenome.Mmusculus.UCSC.mm10[["chr19"]]
chr19_Seq
```

```
##   61431566-letter "DNAString" instance
## seq: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
```

---

# Biostrings and DNAstring

The returned sequence object is a **DNAstring** object from the **Biostrings** package. DNAstrings are efficient objects for storing and accessing large sequences.


```r
class(chr19_Seq)
```

```
## [1] "DNAString"
## attr(,"package")
## [1] "Biostrings"
```

**?DNAstring** for full help on DNAstring objects.
---

# Subsetting a DNAstring set 

We can subset a DNAstring object just like indexing a normal vector


```r
chr19_Seq[1:10000000]
```

```
##   10000000-letter "DNAString" instance
## seq: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...GACCCAACTTGCCCAGGTGCGGGTTTTCCATTG
```

---

# Subsetting a DNAstring set 

Or we can use the Biostrings function **subseq()** on our DNAstring.

Note that **subseq()** is a faster implementation of subsetting for DNAstring objects.


```r
subseq(chr19_Seq,start=1,end=10000000)
```

```
##   10000000-letter "DNAString" instance
## seq: NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...GACCCAACTTGCCCAGGTGCGGGTTTTCCATTG
```

---

# Useful functions.

The Biostrings package has many useful functions for handling sequence data.

We can review the frequency of bases using the **alphabetFrequency()** function.


```r
alphabetFrequency(chr19_Seq)
```

```
##        A        C        G        T        M        R        W        S 
## 16732680 12449343 12420880 16602953        0        0        0        0 
##        Y        K        V        H        D        B        N        - 
##        0        0        0        0        0        0  3225710        0 
##        +        . 
##        0        0
```

---

# Useful functions.

We can get the complement,reverse and reverse complement of sequence using the **complement()**, **reverse()**, **reverseComplement()**.


```r
chr19_SeqComp &lt;- complement(chr19_Seq)
chr19_SeqRev &lt;- reverse(chr19_Seq)
chr19_SeqRevComp &lt;- reverseComplement(chr19_Seq[10000000:10000010])
chr19_Seq[10000000:10000010]
```

```
##   11-letter "DNAString" instance
## seq: GTAATGTACAG
```

```r
chr19_SeqRevComp
```

```
##   11-letter "DNAString" instance
## seq: CTGTACATTAC
```

---

# Useful functions.

We can get also match and count the occurrence of patterns in our sequence using **matchPattern()** and **countPattern()**.

We will look more in depth at these later in the course.


```r
chr19_Count &lt;- countPattern(pattern="ATCTGCAATG",chr19_Seq)
chr19_Count
```

```
## [1] 86
```

---

#Writing to a FASTA file.

The **Biostrings** package contains useful functions to read and write to a FASTA file.

To write our DNAString object to a FASTA file we can use the **writeXStringSet()** function. 

First we will convert our DNAString to a DNAStringSet object using the **DNAStringSet()** function and set names using the **names()** function.



```r
chr19_SeqSet &lt;- DNAStringSet(chr19_Seq[10000000:10000010])
names(chr19_SeqSet) &lt;- "chr19"
writeXStringSet(chr19_SeqSet,filepath = "Data/chr19_Seq.fa")
```



---

#Reading a FASTA file.

Now we can read our a FASTA file back to DNAStringSet object using the **readDNAStringSet()** function. 

We can access our DNAString from the DNAStringSet object again using **$** and **[[]]** functions.



```r
chr19_FromFASTA &lt;- readDNAStringSet(filepath = "Data/chr19_Seq.fa")
chr19_FromFASTA$chr19
```


```
##   11-letter "DNAString" instance
## seq: GTAATGTACAG
```


---

# Time for an exercise.

[Link_to_exercises](../../exercises/fastaAndBiostrings_exercises.html)

[Link_to_answers](../../answers/fastaAndBiostrings_answers.html)
    </textarea>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function() {
  var d = document, s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})();</script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
  }
});
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
